man/man1/xswconv.1
.\" Source file covered by the GNU Public License v3 
.\" See LICENSE file or <http://www.gnu.org/licenses/gpl-3.0.txt/>

.\" Process with
.\" groff -man -Tascii xswconv.1
.\"

.TH XSWCONV 1 "April 2009" Linux "User Manuals"
.SH NAME
xswconv \- converts a XSW presentation into another format

.SH SYNOPSIS
.B xswconv [OPTIONS] -f FORMAT FILE
.br
.B xsw2pdf [OPTIONS] FILE

.br
.SH DESCRIPTION
xswconv is a program that converts a XSW presentation into other formats. It
requires ImageMagick to be installed.

.SH OPTIONS
.TP
\fB-f\fR, \fB\-\-format\fR
choose the output format (see FORMATS)
\fB-o\fR, \fB\-\-output\fR
choose the output filename or directory name (depending of the output file)
.TP
\fB-v\fR, \fB\-\-version\fR
display application version information
.TP
\fB-h\fR, \fB\-\-help\fR
display application help

.SH FORMATS
.TP
\fBpdf\fR
Portable Document Format (pdf) - one file generated per presentation
.TP
\fBpng\fR
Portable Network Graphics (png) - one file generated per slide
.TP
\fBjpeg\fR
JPEG Interchange Format - one file generated per slide

.SH OTHER PROGRAMS
The scripts \fBxsw2pdf\fR, \fBxsw2png\fR and \fBxsw2jpeg\fR are convenience
programs that call \fBxswconv\fR with a predefined output format.

.SH AUTHOR
Andre' Wagner <andre.nho@gmail.com>.
.PP
All development is done at http://xsw.googlecode.com/
.PP
This manual page was written by Andre' Wagner <andre.nho@gmail.com>.

.SH COPYRIGHT
Copyright  (C)  2009 Andre' Wagner
.PP
There is NO warranty.
You may redistribute this software under the terms of  the  GNU
General  Public License.  For more information about these matters, see
the files named COPYING.


NEWS
Version 0.3.5
-------------
* a few files were missing in the last version, and this version corrects
this.

Version 0.3.4
-------------
* created the xswconv, that converts a xsw presentation to png, jpg and pdf.
* created the xsw2pdf, xsw2png and xsw2jpeg, that uses xswconv to convert
a xsw presentation.
* added a -s option to xsw, where the user can start the presentation
at a specific slide.
* added a window icon.
* when xsw enters developer mode in fullscreen, the mouse is now visible.
* corrected a bug where big presentations make xsw enter a race condition
* corrected a bug that prevented developer mode to work.
* corrected a bug that prevented gradients to work when the amount of red.
of the background and the gradient was the same.
* corrected a bug that increased CPU usage in developer mode.

Version 0.3.3
-------------
* xsw no longer install the Vera fonts if they can be found in the font
directory.
* corrected a problem where the font border would be broken if the text
was to fall out of the screen.
* error in configure.ac where ./configure was informing that SDL_image was
missing if SDL_gfx was actually missing.
* added xsw2pdf manual page.
* Changelog file replaced by the NEWS file.

Version 0.3.2
-------------
* created a tutorial presentation, that can be run with the switch --tutorial.
This presentation supersedes the previous introduction presentation.
* corrected a problem where many big presentations would fail due to not
freeing a font.

Version 0.3.1
-------------

* ALT+F4 and CTRL+C now quit xsw, and ESC leaves full screen.
* corrected a bug where image_path didn't accept static paths.
* corrected a bug where xsw2pdf would segfault when the xsw presentation file
didn't have a extension

Version 0.3.0
-------------

* the program 'xsw2pdf' was created. This programs requires ImageMagick and
generates a PDF file from a presentation.
* the color search function was optimized and should be both faster to execute
and to compile.
* the mouse pointer is now hidden in fullscreen.
* a strange number was being shown when CTRL was pressed in developer mode.

Version 0.2.3
-------------

* added a switch '-f' for fullscreen
* the window title now show the current and total slide number

Version 0.2.2
-------------

* named colors: now colors can be referred by their X11 names.
* IDs (template names and custom commands) and named colors are no longer case
sensitive.
* corrected a bug where the background color was not working when using
templates
* added license information to source files

Version 0.2.1
-------------

* fixed a serious bug where previous presentations would not work.

Version 0.2.0
-------------

* added options color and gradient to the background.
* added options color and border to the text.
* added option h (height) to the text.

Version 0.1.3
-------------

* a bug was segfaulting xsw when a empty string (or spaces) was about to be
shown. This was corrected, and now a empty string can be used as a line
spacer.
* an option '--intro' was added to show a introductory slideshow.
* the presentation now can be started from any directory - the program will
look for the images starting on the presentation path.
* the file INSTALL was overwritten by autoconf and was recovered.
* in developer mode, now xsw shows the fine position of the mouse in a box at
the bottom right.

Version 0.1.2
-------------

* the keyword "text" is no longer needed, and it'll be ignored in new versions
* the limit of 1024 slides/commands was removed. There's no limit now.
* the keywords "image" and "background" can be used as "img" and "bg"
* a "-v" option was added the show the program version
* a syntax error now shows the line where the error ocurred
* a option called "image_path" was added, so that xsw will look for images in
that directory as well
* documentation was corrected to inform about comments and keys that can be
used during the execution of the program

Version 0.1.1
-------------

* corrected a problem where the compilation would fail if one of the libraries
was installed in /usr/local.

Version 0.1
-----------

* first public release



trunk/src/xswconv.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <sys/stat.h>
#include "presentation.h"
#include "presenter.h"
#include "slide.h"
#include "file.h"
#include "presenter.h"
#include "parser.h"

extern int parser_parse(Presentation *pres, char *filename);


// 
// available formats
//
typedef enum { none, pdf, png, jpeg } t_format;
typedef enum { one_file_per_slide, one_file_per_presentation } t_nfiles;

typedef struct {
	char* desc;
	t_format format;
	char* extension;
	t_nfiles n_files;
} t_type;

static const t_type const types[] = {
	{ "none", none, "none", one_file_per_slide },
	{ "pdf", pdf, "pdf", one_file_per_presentation },
	{ "png", png, "png", one_file_per_slide },
	{ "jpeg", jpeg, "jpg", one_file_per_slide }
};

#define TYPE_NONE types[0];
#define N_TYPES (sizeof(types) / sizeof(*types))

// 
// user options
//
typedef struct {
	t_type type;
	char* output_file;
	char* filename;
} ConvOptions;


// 
// check if the utility 'convert' is installed
//
static int check_convert()
{
	if(system("convert > /dev/null") == 32512)
		return 0;
	return 1;
}

// 
// print tool usage
//
static void print_usage(FILE* stream, int exit_value)
{
	fprintf(stream, "xswconv version " VERSION "\n");
	fprintf(stream, "Usage: xswconv [OPTIONS] FILE\n");
	fprintf(stream, "Convert a xsw presentation into a different format.\n");
	fprintf(stream, "\n");
	fprintf(stream, "Options:\n");
	fprintf(stream, "  -f, --format=FORMAT     output format (see below)\n");
	fprintf(stream, "  -o, --output=FILE       output file/directory name\n");
	fprintf(stream, "\n");
	fprintf(stream, "Formats:\n");
	fprintf(stream, "   pdf       Portable Document Format (one file per presentation)\n");
	fprintf(stream, "   png       Portable Network Graphics (one file per slide)\n");
	fprintf(stream, "  jpeg       JPEG interchange format (one file per slide)\n");

	exit(exit_value);
}


// 
// print tool version
//
static void print_version(FILE* stream, int exit_code)
{
	fprintf(stream, "xswconv " VERSION ", a xsw converter tool.\n");
	fprintf(stream, "xsw and xswconv are free softwares.\n");
	fprintf(stream, "Copyleft (C) 2009 André Wagner <andre.nho@gmail.com> - all wrongs reserved\n");
	exit(exit_code);
}


// 
// parse the options given by the user
//
static ConvOptions* parse_options(int argc, char** argv)
{
	int next_option;
	unsigned int i;

	ConvOptions* opt = malloc(sizeof(ConvOptions));
	opt->type = TYPE_NONE;
	opt->output_file = NULL;
	opt->filename = NULL;

	const char* const short_options = "hvf:o:";
	const struct option long_options[] = {
		{ "help", 0, NULL, 'h' },
		{ "version", 0, NULL, 'v' },
		{ "format", 1, NULL, 'f' },
		{ "output", 1, NULL, 'o' },
		{ NULL, 0, NULL, 0 }
	};

	do 
	{
		next_option = getopt_long(argc, argv, short_options, 
				long_options, NULL);
		switch(next_option)
		{
			case '?':
				print_usage(stderr, 1);
				break;
			case -1: // done
				break;
			case 'h':
				print_usage(stdout, 0);
			case 'v':
				print_version(stdout, 0);
			case 'f':
				for(i=0; i<N_TYPES; i++)
					if(strcmp(optarg, types[i].desc) == 0)
						opt->type = types[i];
				break;
			case 'o':
				opt->output_file = strdup(optarg);
				break;
			default:
				abort();
		}
	} while (next_option != -1);

	if(argc > optind)
	{
		opt->filename = strdup(argv[optind]);

		char *s = strdup(opt->filename);
		if(!opt->output_file)
			opt->output_file = base_name(s);
	}

	return opt;
}


// 
// parse options and generate the files
//
int main(int argc, char** argv)
{
	// check parameters
	ConvOptions *opt = parse_options(argc, argv);
	if(opt->type.format == none || !opt->output_file || !opt->filename)
		print_usage(stderr, 1);

	// check for 'convert'
	if(!check_convert())
	{
		fprintf(stderr, "Sorry, the 'convert' program was not found. It is part of the ImageMagick\ntools. You can get it at: <http://www.imagemagick.org/>\n");
		return 1;
	}

	// load presentation
	Presentation* p = presentation_new();
	p->filename = opt->filename;
	if(!file_exists(p->filename))
	{
		fprintf(stderr, "%s is not a valid file.\n", p->filename);
		return 1;
	}
	p->path = file_path(p->filename);

	// parse code in the file
	if(!parser_parse(p, p->filename))
	{
		fprintf(stderr, "Invalid XSW file.\n");
		return 1;
	}

	// prepare temporary directory
	char* path = alloca(20);
	if (path == NULL) 
	{
		fprintf(stderr, "Not enough memory\n");
		exit(1);		
	}
	sprintf(path, "/tmp/%s", "xsw-XXXXXX");
	path = mkdtemp(path);

	// load slideshow
	Presenter* pr = presenter_initialize(p, 0);
	int slides = lcount(p->slides);
	int i;
	char* all_bmp = alloca(slides * 25); // TODO
	if (all_bmp == NULL) 
	{
		fprintf(stderr, "Not enough memory\n");
		exit(1);		
	}
	strcpy(all_bmp, "");
	for(i=0; i<slides; i++)
	{
		char bmp_path[512];
		sprintf(bmp_path, "%s/%d.bmp", path, i);
		presenter_show(pr, i, 0);
		presenter_save_image(pr, bmp_path);
		strcat(all_bmp, bmp_path);
		strcat(all_bmp, " ");
	}

	// generate output file
	if(opt->type.n_files == one_file_per_presentation)
	{
		char* command = alloca(strlen(all_bmp) + 512);
		if (command == NULL) 
		{
			fprintf(stderr, "Not enough memory\n");
			exit(1);		
		}
		char* name = alloca(512);
		if (name == NULL) 
		{
			fprintf(stderr, "Not enough memory\n");
			exit(1);		
		}
		sprintf(name, "%s.pdf", base_name(p->filename));
		printf("Generating %s...\n", name);
		sprintf(command, "convert -compress Zip %s %s", all_bmp, name);
		if(system(command) != 0)
		{
			fprintf(stderr, "There was an error when running 'convert'.\n");
			return 1;
		}
	}
	else
	{
		// make directory
		if(mkdir(opt->output_file, S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IXGRP|S_IROTH|S_IXOTH) != 0)
		{
			fprintf(stderr, "Could not create directory %s/ .\n", opt->output_file);
			return 1;
		}

		// convert files
		for(i=0; i<slides; i++)
		{
			char* command = alloca(1024);
			sprintf(command, "convert %s/%d.bmp %s/%d.%s",
					path, i, opt->output_file, i, opt->type.extension);
			printf("Generating slide %d.\n", i+1);
			if(system(command) != 0)
			{
				fprintf(stderr, "There was an error when running 'convert'.\n");
				return 1;
			}
		}
	}

	// remove files
	printf("Removing temporary files...\n");
	for(i=0; i<slides; i++)
	{
		char bmp_path[512];
		sprintf(bmp_path, "%s/%d.bmp", path, i);
		unlink(bmp_path);
	}
	rmdir(path);

	return 0;
}
